File: src/correlate/linker.js
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const CORR_CONFIG_PATH = join(__dirname, '..', '..', 'config', 'correlation.json');

export class Correlator {
    constructor(state) {
        this.state = state;
        this.config = JSON.parse(readFileSync(CORR_CONFIG_PATH, 'utf-8'));
    }

    correlate(investigationId) {
        const entities = this.state.getEntities(investigationId);
        const existingLinks = new Set(
            this.state.getLinks(investigationId).map(l => `${l.from_entity_id}:${l.to_entity_id}:${l.link_type}`)
        );

        for (const rule of this.config.linking_rules) {
            const fromEntities = entities.filter(e => e.type === rule.from);
            const toEntities = entities.filter(e => e.type === rule.to);

            for (const from of fromEntities) {
                for (const to of toEntities) {
                    if (from.id === to.id) continue;

                    const confidence = this._matchScore(from, to, rule);
                    if (confidence < (this.config.scoring?.fuzzy_match_threshold || 0.5)) continue;

                    const linkKey = `${from.id}:${to.id}:${rule.name}`;
                    const reverseLinkKey = `${to.id}:${from.id}:${rule.name}`;
                    if (existingLinks.has(linkKey) || existingLinks.has(reverseLinkKey)) continue;

                    this.state.addLink(investigationId, from.id, to.id, rule.name, confidence, [
                        { rule: rule.name, from_type: from.type, to_type: to.type }
                    ]);
                    existingLinks.add(linkKey);

                    if (rule.bidirectional) {
                        this.state.addLink(investigationId, to.id, from.id, rule.name, confidence, [
                            { rule: rule.name, from_type: to.type, to_type: from.type, reverse: true }
                        ]);
                        existingLinks.add(reverseLinkKey);
                    }
                }
            }
        }

        this._applyMultiSourceBonus(investigationId, entities);
        return this.state.getLinks(investigationId);
    }

    _matchScore(from, to, rule) {
        let matched = 0;
        let total = rule.match_fields.length;

        for (const field of rule.match_fields) {
            const fromVal = this._resolveField(from, field);
            const toVal = this._resolveField(to, field);

            if (!fromVal || !toVal) continue;

            if (typeof fromVal === 'string' && typeof toVal === 'string') {
                if (fromVal.toLowerCase() === toVal.toLowerCase()) {
                    matched++;
                } else if (this._fuzzyMatch(fromVal, toVal) > 0.8) {
                    matched += 0.7;
                }
            } else if (fromVal === toVal) {
                matched++;
            }
        }

        if (total === 0) return 0;
        const matchRatio = matched / total;
        return matchRatio * rule.confidence;
    }

    _resolveField(entity, field) {
        if (entity.data[field] !== undefined) return entity.data[field];

        const aliases = {
            address: ['email', 'ip_address', 'ip'],
            ip: ['address', 'ip_address'],
            name: ['domain', 'hostname', 'subdomain', 'login', 'title'],
            username: ['handle', 'screen_name', 'user', 'login'],
            number: ['phone', 'phone_number'],
            email: ['address'],
            homepage: ['url', 'blog']
        };

        for (const [canonical, alts] of Object.entries(aliases)) {
            if (field === canonical) {
                for (const alt of alts) {
                    if (entity.data[alt] !== undefined) return entity.data[alt];
                }
            }
        }

        if (field === 'username' && entity.type === 'repository' && entity.data.name) {
            const parts = entity.data.name.split('/');
            if (parts.length === 2) return parts[0];
        }

        if (field === 'name' && entity.type === 'url' && entity.data.url) {
            try {
                return new URL(entity.data.url.startsWith('http') ? entity.data.url : `https://${entity.data.url}`).hostname;
            } catch {}
        }

        if (field === 'name' && entity.type === 'technology' && entity.data.url) {
            try {
                return new URL(entity.data.url.startsWith('http') ? entity.data.url : `https://${entity.data.url}`).hostname;
            } catch {}
        }

        return null;
    }

    _fuzzyMatch(a, b) {
        const la = a.toLowerCase();
        const lb = b.toLowerCase();
        if (la === lb) return 1.0;
        if (la.includes(lb) || lb.includes(la)) return 0.9;

        const maxLen = Math.max(la.length, lb.length);
        if (maxLen === 0) return 1.0;
        const dist = this._levenshtein(la, lb);
        return 1.0 - (dist / maxLen);
    }

    _levenshtein(a, b) {
        const m = a.length, n = b.length;
        const dp = Array.from({ length: m + 1 }, (_, i) => {
            const row = new Array(n + 1);
            row[0] = i;
            return row;
        });
        for (let j = 0; j <= n; j++) dp[0][j] = j;

        for (let i = 1; i <= m; i++) {
            for (let j = 1; j <= n; j++) {
                dp[i][j] = a[i - 1] === b[j - 1]
                    ? dp[i - 1][j - 1]
                    : 1 + Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]);
            }
        }
        return dp[m][n];
    }

    _applyMultiSourceBonus(investigationId, entities) {
        const bonus = this.config.scoring?.multi_source_bonus || 0.1;
        for (const entity of entities) {
            if (entity.source_count > 1) {
                const boosted = Math.min(1.0, entity.confidence + (bonus * (entity.source_count - 1)));
                if (boosted !== entity.confidence) {
                    this.state.db.prepare(
                        'UPDATE entities SET confidence = ? WHERE id = ?'
                    ).run(boosted, entity.id);
                }
            }
        }
    }

    buildGraph(investigationId) {
        const entities = this.state.getEntities(investigationId);
        const links = this.state.getLinks(investigationId);

        return {
            nodes: entities.map(e => ({
                id: e.id,
                type: e.type,
                label: this._entityLabel(e),
                confidence: e.confidence,
                sources: e.source_count,
                data: e.data
            })),
            edges: links.map(l => ({
                from: l.from_entity_id,
                to: l.to_entity_id,
                type: l.link_type,
                confidence: l.confidence
            }))
        };
    }

    _entityLabel(entity) {
        const d = entity.data;
        switch (entity.type) {
            case 'person': return d.name || d.username || d.email || 'Unknown Person';
            case 'account': return `${d.platform}/${d.username}`;
            case 'email': return d.address;
            case 'domain': return d.name;
            case 'subdomain': return d.name;
            case 'ip': return d.address;
            case 'port': return `${d.address}:${d.number}`;
            case 'breach': return d.name;
            case 'credential': return `${d.email}@${d.source_breach}`;
            case 'phone': return d.number;
            case 'url': return d.url;
            case 'certificate': return d.subject || d.fingerprint;
            case 'dns_record': return `${d.type} ${d.name}`;
            default: return JSON.stringify(d).slice(0, 50);
        }
    }
}

File: cyclops/correlate/linker.py
import json
from pathlib import Path

CONFIG_PATH = Path(__file__).parent.parent.parent / "config" / "correlation.json"


def levenshtein(a, b):
    if len(a) < len(b):
        return levenshtein(b, a)
    if not b:
        return len(a)
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a):
        curr = [i + 1]
        for j, cb in enumerate(b):
            curr.append(min(prev[j + 1] + 1, curr[j] + 1, prev[j] + (0 if ca == cb else 1)))
        prev = curr
    return prev[-1]


def similarity(a, b):
    if not a or not b:
        return 0.0
    dist = levenshtein(a.lower(), b.lower())
    max_len = max(len(a), len(b))
    return 1 - (dist / max_len)


class Correlator:
    def __init__(self, state, telemetry):
        self.state = state
        self.telemetry = telemetry
        self.config = json.loads(CONFIG_PATH.read_text()) if CONFIG_PATH.exists() else {}
        self.threshold = self.config.get("fuzzy_threshold", 0.85)

    def correlate(self, inv_id):
        entities = self.state.get_entities(inv_id)
        links = []

        for i, e1 in enumerate(entities):
            for e2 in entities[i + 1:]:
                link = self._check_link(e1, e2)
                if link:
                    self.state.add_link(inv_id, e1["id"], e2["id"], link["type"], link["confidence"], link.get("evidence"))
                    links.append(link)

        self._apply_multi_source_bonus(inv_id)
        return links

    def _check_link(self, e1, e2):
        d1, d2 = e1["data"], e2["data"]

        if e1["type"] == "email" and e2["type"] == "account":
            if d1.get("address") and d1["address"] == d2.get("address"):
                return {"type": "email_to_account", "confidence": 0.95, "evidence": [{"rule": "exact_email_match"}]}

        if e1["type"] == "account" and e2["type"] == "email":
            if d1.get("address") and d1["address"] == d2.get("address"):
                return {"type": "account_to_email", "confidence": 0.95, "evidence": [{"rule": "exact_email_match"}]}

        if e1["type"] == "email" and e2["type"] == "breach":
            if d1.get("address") and d1["address"] == d2.get("address"):
                return {"type": "email_in_breach", "confidence": 0.9, "evidence": [{"rule": "email_breach_match"}]}

        if e1["type"] == "person" and e2["type"] == "account":
            u1 = d1.get("username", "")
            u2 = d2.get("username", "")
            if u1 and u2:
                sim = similarity(u1, u2)
                if sim >= self.threshold:
                    return {"type": "person_to_account", "confidence": sim, "evidence": [{"rule": "username_fuzzy", "similarity": sim}]}

        if e1["type"] == "account" and e2["type"] == "account":
            u1 = d1.get("username", "")
            u2 = d2.get("username", "")
            if u1 and u2 and d1.get("platform") != d2.get("platform"):
                sim = similarity(u1, u2)
                if sim >= self.threshold:
                    return {"type": "cross_platform", "confidence": sim * 0.8, "evidence": [{"rule": "username_fuzzy", "similarity": sim}]}

        if e1["type"] == "domain" and e2["type"] == "ip":
            return None
        if e1["type"] == "ip" and e2["type"] == "domain":
            return None

        if e1["type"] == "subdomain" and e2["type"] == "domain":
            if d1.get("name", "").endswith(f".{d2.get('name', '')}"):
                return {"type": "subdomain_of", "confidence": 1.0, "evidence": [{"rule": "parent_domain"}]}

        if e1["type"] == "domain" and e2["type"] == "subdomain":
            if d2.get("name", "").endswith(f".{d1.get('name', '')}"):
                return {"type": "has_subdomain", "confidence": 1.0, "evidence": [{"rule": "parent_domain"}]}

        return None

    def _apply_multi_source_bonus(self, inv_id):
        entities = self.state.get_entities(inv_id)
        for e in entities:
            if e["source_count"] >= 3 and e["confidence"] < 0.9:
                bonus = min(1.0, e["confidence"] + 0.1)
                self.state.db.execute("UPDATE entities SET confidence = ? WHERE id = ?", (bonus, e["id"]))
        self.state.db.commit()

    def build_graph(self, inv_id):
        entities = self.state.get_entities(inv_id)
        links = self.state.get_links(inv_id)

        nodes = []
        for e in entities:
            nodes.append({
                "id": e["id"],
                "type": e["type"],
                "label": self._entity_label(e),
                "confidence": e["confidence"],
            })

        edges = []
        for l in links:
            edges.append({
                "from": l["source_id"],
                "to": l["target_id"],
                "type": l["type"],
                "confidence": l["confidence"],
            })

        return {"nodes": nodes, "edges": edges}

    def _entity_label(self, entity):
        d = entity["data"]
        t = entity["type"]
        labels = {
            "email": lambda: d.get("address", ""),
            "account": lambda: f"{d.get('platform', '')}:{d.get('username', '')}",
            "person": lambda: d.get("name") or d.get("username", ""),
            "domain": lambda: d.get("name", ""),
            "subdomain": lambda: d.get("name", ""),
            "ip": lambda: d.get("address", ""),
            "port": lambda: f"{d.get('ip', '')}:{d.get('number', '')}",
            "certificate": lambda: d.get("subject", ""),
            "breach": lambda: d.get("name", ""),
            "phone": lambda: d.get("number", ""),
            "url": lambda: d.get("url", ""),
            "dns_record": lambda: f"{d.get('type', '')} {d.get('value', '')}",
            "repository": lambda: d.get("name") or d.get("url", ""),
            "organization": lambda: d.get("name", ""),
            "technology": lambda: ", ".join(d.get("technologies", [])) if d.get("technologies") else d.get("url", ""),
        }
        return labels.get(t, lambda: str(d))()
